### PR TITLE
fix: ConnectionMiddleware and IEnumerable + IConnection

### PR DESCRIPTION
**Pull request by benmccallum**
Don't assume an `IEnumerable`/`IQueryable` isn't already an `IConnection`

### CODE REVIEW

**Comment by benmccallum** [comment]
Disclaimer: I haven't tested this at all, but assuming all automated tests pass it *could* be ok :P

**Comment by michaelstaib** [comment]
We should write one test to prove that it works.

**Review by michaelstaib — DISMISSED** [review]

File: src/Core/Types/Types/Relay/ConnectionMiddleware.cs [hunk 1]
```diff
@@ -30,7 +30,11 @@ public ConnectionMiddleware(FieldDelegate next)
                 context.Argument<string>("after"),
                 context.Argument<string>("before"));
 
-            if (connectionResolver is { } && context.Result is TSource source)
+            if (context.Result is IConnection)
```
Comment (michaelstaib) [inline src/Core/Types/Types/Relay/ConnectionMiddleware.cs:33]: I will have to look at the new implementation so that this will work with 11 as well.

**Comment by benmccallum** [comment]
Added a unit test for this scenario, but I'm not too clued up on the other scenarios, which don't currently have unit tests, so perhaps you'd want to quickly bang out some of those for the other types the middleware can handle (like IQueryable, IEnumerable that's not an `IConnection`, etc.)?

I'm just wondering one thing. Should the `IConnection` scenario check inside to make sure that the `Edge<T>`, that `T` is `TEntity`? I'm thinking not, as it's more flexibile. Essentially we're saying if you're returning an `IConnection`, you know what you're doing and we'll take it as is. e.g. in my case I'm returning an `IConnection` of `Supplier`, but the field is typed as a connection of `SupplierResult` (a union that includes supplier) as I have middleware that swaps out supplier's for `GoneSupplier` when the supplier has left our platform as the user isn't an admin.

File: src/Core/Types.Tests/Types/Relay/__snapshots__/ConnectionMiddlewareTests.ExecuteQueryWithConnectionMiddleware_ShouldHandleVariousReturnTypes.snap [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      ]
+    },
+    "foosAsIEnumerable": {
+      "totalCount": 4,
+      "pageInfo": {
+        "hasNextPage": false,
+        "hasPreviousPage": false,
+        "startCursor": "MA==",
+        "endCursor": "Mw=="
+      },
```
Comment (benmccallum) [inline src/Core/Types.Tests/Types/Relay/__snapshots__/ConnectionMiddlewareTests.ExecuteQueryWithConnectionMiddleware_ShouldHandleVariousReturnTypes.snap:37]: ^^ Is wrong.

File: src/Core/Types.Tests/Types/Relay/__snapshots__/ConnectionMiddlewareTests.ExecuteQueryWithConnectionMiddleware_ShouldHandleVariousReturnTypes.snap [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      ]
+    },
+    "foosAsIQueryable": {
+      "totalCount": 4,
+      "pageInfo": {
+        "hasNextPage": false,
+        "hasPreviousPage": false,
+        "startCursor": "MA==",
+        "endCursor": "Mw=="
+      },
```
Comment (benmccallum) [inline src/Core/Types.Tests/Types/Relay/__snapshots__/ConnectionMiddlewareTests.ExecuteQueryWithConnectionMiddleware_ShouldHandleVariousReturnTypes.snap:64]: ^^ Is wrong.

**Comment by benmccallum** [comment]
Added a basic IQueryable and IEnumerable test.

**Review by michaelstaib — APPROVED** [review]

**Pull request merged**